NewApp/ngsa-csharp/Ngsa.Middleware/NgsaLogger/NgsaLogger.cs

```diff
@@ -0,0 +1,181 @@
+            {
+                CorrelationVector cv = CorrelationVectorExtensions.GetCorrelationVectorFromContext(context);
+
+                if (cv != null)
+                {
+                    d.Add("CVector", cv.Value);
+                }
+            }
+
+            if (data != null)
+            {
+                foreach (var kvp in data)
+                {
+                    d.Add(kvp.Key, kvp.Value);
+                }
+            }
+
+            // display the error
+            Console.ForegroundColor = ConsoleColor.Red;
+            Console.Error.WriteLine(JsonSerializer.Serialize(d));
+            Console.ForegroundColor = origColor;
+
+            // free the memory for GC
+            d.Clear();
+        }
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            if (!IsEnabled(logLevel))
+            {
+                return;
+            }
+
+            Dictionary<string, object> d = new Dictionary<string, object>
+            {
+                { "logName", name },
+                { "logLevel", logLevel.ToString() },
+                { "eventId", eventId.Id },
+                { "eventName", eventId.Name },
+            };
+
+            // convert state to list
+            if (state is IReadOnlyList<KeyValuePair<string, object>> roList)
+            {
+                List<KeyValuePair<string, object>> list = roList.ToList();
+
+                switch (list.Count)
+                {
+                    case 0:
+                        break;
+                    case 1:
+                        // clean up name
+                        list.Add(new KeyValuePair<string, object>("message", list[0].Value));
+                        list.RemoveAt(0);
+                        break;
+                    default:
+                        // remove formatting key-value
+                        list.RemoveAt(list.Count - 1);
+                        break;
+                }
+
+                for (int i = list.Count - 1; i >= 0; i--)
+                {
+                    // get correlation vector from HttpContext.Items
+                    if (list[i].Value is HttpContext c)
+                    {
+                        list.RemoveAt(i);
+
+                        if (c != null && c.Items != null)
+                        {
+                            CorrelationVector cv = CorrelationVectorExtensions.GetCorrelationVectorFromContext(c);
+
+                            if (cv != null)
+                            {
+                                d.Add("CVector", cv.Value);
+                            }
+                        }
+
+                        break;
+                    }
+                }
+
+                // add remaining state
+                foreach (var kvp in list)
+                {
+                    d.Add(kvp.Key.ToString(), kvp.Value.ToString());
+                }
```

Comment:
This looks like you might loop the list almost fully twice if the HttpContext object was at the end of the list.  Could we just do it in 1 loop and if it is not a context then add to the kvp dict?  I would mean more conditionals execute, but also would mean that we don't need to remove it from the list.

Reply:
fixed